EuiComboBox

cjcenizal opened the pull request:
Replaces https://github.com/elastic/eui/pull/52

### Input

- [x] When the user start typing into the input, it should move to the next line if necessary to give the user as much room to type as possible.

### Options list

- [x] Options should highlight substrings which match the search input.
- [x] Consumer needs to be able to group options.
- [x] Nice-to-have: Options should be custom-renderable.
- [x] The list should render inside a portal, and reposition itself when it's about to flow off-screen.

### Use cases

- [x] Multi-select (http://jedwatson.github.io/react-select/)
- [x] Nice-to-have: Single-select (http://uniform.hudl.com/components/selects/lookup-select/design)
- [x] Nice-to-have: Free-form input AKA discrete value input (https://github.com/elastic/eui/pull/353)

### Empty states

- [x] When nothing matches the input, suggest adding it by hitting enter.
- [x] When there are no suggestions remaining, surface some appropriate feedback, e.g. "You've selected all available options".

### Keyboard accessibility

- [x] Tabbing to the component will focus the input
- [x] When the cursor is at the beginning of the input:
  - [x] hitting backspace will delete the preceding pill
  - [x] ~~hitting left arrow will focus that pill~~
- [x] ~~Left and right arrow keys can be used to navigate through the pills~~
- [x] Up and down arrow keys can be used to navigate into, throughout, and out of the options list

### Screen reader accessibility

- [x] Possible actions will be announced at each stage of the above
- [x] The value of the input will contain the selected values; we'll use `aria-label` to read the humanized values aloud

### Async actions

- [x] We'll want to be able to display a loading spinner while it's waiting on options to be loaded (`isLoading` prop)
- [x] We'll also want the ability to save a new pill value

cjcenizal commented:
@nreese @snide @peteharverson Could you please take a look at how this behaves in the browser and give me your thoughts before I add tests? @timroes Could you do the same, especially regarding screen reader and keyboard accessibility? Thanks!

snide commented:
This is great and covers almost all the use cases I can think of (minus the single select one, which I see you have as another item). I'll have some styling to refine it, but all the bits are there so it should be a T-Ball. Big thumbs up. The following are some finesse suggestions.

1. We need a placeholder prop. It can just place itself in the "input". I can see instances where the placeholder might be different if there are or are not pills already set, so if you happen to set up two props for it, that would be pretty awesome.
2. I think we should separate the up/down selection from the input focus. Right now you can't move up and down and type at the same time. I think you want to keep the cursor focused in the input as they move up and down. I know that means you don't get the benefit of auto-scrolling because of the item focus, but having them as separate actions means your hands need to do much less work to move around. With the aria stuff I think you'd have to read the selection independently, but you can see the benefit there too, since it will read the first match as you type.
3. Tied to that, I'd auto-select the first option so you can just jam on enter immediately. As mentioned before, should also help speed up a screen reader's time to action as well.
4. I think the "pill" should be tabable. And that you should have to tab through them to get to the input. Right now it's pretty hard to get to them with your keyboard. There's no way to delete the first item of three with the keyboard. You have to delete all or nothing.
5. Similarly, for your backspace mechanic, I'd suggest using the first Backspace to do a "shift-tab" action, that actually selects the item, then the second backspace actually deletes it. This prevents people from inadvertently deleting things by accident. It's a small thing, but really helps these kinds of interactions. Here's an example of one I built awhile ago. The "blue" state is the first backspace, and then the second backspace confirms it. If you end up added something like that, i can think up some proper styling if you pass over an `isSelected` prop on the pill itself.

![](https://d3vv6lp55qjaqc.cloudfront.net/items/3O1b0Y2O0r2t32372V02/Screen%20Recording%202018-03-27%20at%2010.02%20PM.gif?X-CloudApp-Visitor-Id=59773&v=10450270)

I know this is a big one so I'm sure I'll have other stuff, but those are the bits that jumped out at me immediately. All that said it's really impressive and mostly there, certainly from a functionality scenario.

cjcenizal commented:
@bevacqua This component supports a mode in which the suggestions list is hidden and the user can only type custom values which get turned into pills. It looks like this could possibly replace https://github.com/elastic/eui/pull/353. Can you take a look and let me know if this satisfies your use cases for that component?

timroes commented:
Thanks for creating this. I will try to review (accessibility) as soon as possible.

Single select would btw be a "super-nice-to-have", since that would make it usable in most places where we currently use `react-select` in vis editors (or older Angular components) for all the aggregation selection, etc.

jgowdyelastic commented:
Great progress!
For this to be a complete replacement for ML's use of `ui-select` we would definitely need single selection and the ability to group items under subheadings.
For example, this select in our Multi-metric job creation page.
![image](https://user-images.githubusercontent.com/22172091/38019989-01e25c78-3271-11e8-9c51-7d410fe797b8.png)
Groups isn't a must, but i'm mentioning it because i didn't see it on the feature list

cjcenizal commented:
@bevacqua Thanks for your feedback. It sounds like this component _could_ substitute for some of the use cases of your discrete input component, but it's not quite as flexible as you want. I think it will probably make both development and consumption easier if we pursue the direction you suggest as a separate component.

@timroes @jgowdyelastic I'll implement single-selection before merging. Thanks for pointing out its importance.

@nreese Per our convo I'll also render the options list within a portal, so EuiComboBox can be used inside of dashboards.

@snide Thanks for your comments. To your points:

1. Good idea on the placeholder. I'll implement a placeholder for when there aren't any pills for now, and create an issue for a placeholder when there are pills. I see how you used it in the Guidebook example. I was looking at examples from [Uniform](http://uniform.hudl.com/components/selects/lookup-select/design/) and [react-select](http://jedwatson.github.io/react-select/) and they felt pretty intuitive to me without the persistent placeholder so I wonder if the component is better without it.
2. I think keeping focus on the input as you traverse the options is a really neat idea. I think I know how we can implement this but I don't think I have the time to do it right now, so I'll create an issue for this.
3. Same.
4. Same here. Great idea, something I thought of too.
5. Another great idea. I'm all for protecting against accidental destructive actions. I'll create an issue for it.

cjcenizal commented:
@snide @timroes @jgowdyelastic @nreese I've addressed your comments. Please take another look so I can merge this. I hate to do it but I think tests will have to come separately.

nreese commented:
@cjcenizal I am seeing some weird behavior with the latest version in Kibana. Opening the options list makes the rest of the page content disappear.

![portal_problems](https://user-images.githubusercontent.com/373691/38063362-f3a6c564-32b5-11e8-8a96-d85e767cbcc7.gif)

nreese commented:
@cjcenizal Thanks for the note about the euiBody-hasToolTip rename to euiBody-hasPortalContent. Kibana has a css rule overriding euiBody-hasToolTip position and when that override was applied, the problems above were fixed

nreese commented:
Sometimes the options box is wider than the input.

<img width="571" alt="screen shot 2018-03-29 at 9 14 48 am" src="https://user-images.githubusercontent.com/373691/38097115-bb3f13da-3331-11e8-831f-9d9afc95c6f3.png">

nreese commented on src-docs/src/views/combo_box/single_selection.js:
```diff
@@ -0,0 +1,67 @@
+    };
+  }
+
+  onChange = (selectedOption) => {
```
This example is a little confusing because `selectedOption` is an array. Maybe it would be clearer if its written like

```
onChange = (selectedOptions) => {
  let selectedOption;
  if (selectedOptions.length > 0) {
    selectedOption = selectedOptions[0];
  this.setState({
    selectedOption
  });
}
```

cjcenizal replied:
Will do.

cjcenizal commented:
I think I've run out of time on this. I'm on PTO tomorrow and next week I need to shift my focus to Cloud. Unless anyone object, I'm going to merge this tomorrow morning. Here are the outstanding issues I know of that still need to be addressed:

* There's a TODO in combo_box.js which notes that we need to wait for the EUI CSS to be loaded before calling `this.autoSizeInput.copyInputStyles()` ([docs](https://github.com/JedWatson/react-input-autosize#changing-the-styles-at-runtime). Ideally we should listen for the styles to load instead of using a static timeout. Until we fix this, users are bound to see the input text with the incorrect font-size and style.
* EuiComboBox doesn't work inside of EuiModal. There's a problem with the tab order and the way the ESCAPE keypress is handled.
* Nate's sizing issue hasn't been addressed yet. We need to intelligently set the width of the options list to match that of the input box.

Once this is merged I'll break out individual issues for these.

cjcenizal commented:
Oh and thanks again for the rigorous testing and great feedback everyone. :)

nreese commented:
There is a problem with `addCustomOption` and `doesSearchMatchOnlyOption`. `doesSearchMatchOnlyOption` checks for one remaining option on `this.matchingOptions`. After `doesSearchMatchOnlyOption` returns true, `addCustomOption` assumes that `this.options` has only one option and calls click.
```
  if (this.doesSearchMatchOnlyOption()) {
      this.options[0].click();
      return;
    }
```
However there is a way to enter `addCustomOption` when the options menu is not visible so `this.options` is empty even though `this.matchingOptions` has one item. This is achieved by deleting the last item in the ComboBox. The cursor stays in the input but the options panel is not visible. 

![box](https://user-images.githubusercontent.com/373691/38119401-c1e90f40-337e-11e8-9669-f8b44eb2a8f8.gif)

cjcenizal commented:
I addressed the bug @nreese found. Merging now!
